presto-main/src/main/java/com/facebook/presto/sql/gen/ExpressionCompiler.java

```diff
@@ -277,43 +278,45 @@ else if (expression instanceof LambdaDefinitionExpression) {
     // left to right.
     List<Supplier<PageFilter>> makeReorderableFilters(Optional<RowExpression> filter, Optional<String> classNameSuffix)
     {
-        ArrayList<Supplier<PageFilter>> result = new ArrayList();
         if (!filter.isPresent()) {
-            return result;
+            return ImmutableList.of();
         }
-        ArrayList<RowExpression> conjuncts = new ArrayList();
-        collectConjuncts(filter.get(), conjuncts);
-        if (conjuncts.size() == 1) {
-            result.add(pageFunctionCompiler.compileFilter(filter.get(), classNameSuffix));
+
+        ImmutableList.Builder<RowExpression> conjunctBuilder = ImmutableList.builder();
+        collectConjuncts(filter.get(), conjunctBuilder);
+        List<RowExpression> allConjuncts = conjunctBuilder.build();
+
+        if (allConjuncts.size() == 1) {
+            return ImmutableList.of(pageFunctionCompiler.compileFilter(filter.get(), classNameSuffix));
         }
-        else {
-            CallExpression topAnd = (CallExpression) filter.get();
-            HashMap<HashSet<InputReferenceExpression>, ArrayList<RowExpression>> inputsToConjuncts = new HashMap();
-            for (RowExpression conjunct : conjuncts) {
-                HashSet<InputReferenceExpression> inputs = new HashSet();
-                collectInputs(conjunct, inputs);
-                ArrayList<RowExpression> list = inputsToConjuncts.get(inputs);
-                if (list == null) {
-                    list = new ArrayList();
-                    list.add(conjunct);
-                    inputsToConjuncts.put(inputs, list);
-                }
-                else {
-                    list.add(conjunct);
-                }
-            }
-            for (Map.Entry<HashSet<InputReferenceExpression>, ArrayList<RowExpression>> entry : inputsToConjuncts.entrySet()) {
-                ArrayList<RowExpression> list = entry.getValue();
-                RowExpression conjunct = list.get(0);
-                for (int i = 1; i < list.size(); i++) {
-                    ArrayList<RowExpression> arguments = new ArrayList();
-                    arguments.add(conjunct);
-                    arguments.add(conjuncts.get(i));
-                    conjunct = new CallExpression(topAnd.getSignature(), topAnd.getType(), arguments);
-                }
-                result.add(pageFunctionCompiler.compileFilter(conjunct, classNameSuffix));
+
+        ImmutableList.Builder<Supplier<PageFilter>> result = ImmutableList.builder();
+        CallExpression topAnd = (CallExpression) filter.get();
+        Map<Set<InputReferenceExpression>, List<RowExpression>> inputsToConjuncts = new HashMap();
+        for (RowExpression conjunct : allConjuncts) {
+            ImmutableSet.Builder<InputReferenceExpression> inputs = ImmutableSet.builder();
+            collectInputs(conjunct, inputs);
+            inputsToConjuncts.compute(inputs.build(),
```

Comment:
What do you think of this:
`inputsToConjuncts.computeIfAbsent(inputs.build(), k -> new ArrayList<>).add(conjunct);`